File: bindings/wasm/tests/evm-vectors.test.ts
// Phase 6 EVM vector parity tests for the WASM binding.
//
// Mirrors `bindings/kotlin/.../EvmVectorsTest.kt`: every vector in
// `spec/test-vectors.json` whose chain (for address vectors) or unsigned_tx
// (for sign_tx) is in the EVM family is exercised through the WASM
// `JovaWallet`, asserting byte-identical output against the captured cast
// reference values.

import { describe, it, expect } from 'vitest';

import { JovaWallet, JovaException } from '../src/index.js';
import vectors from '../../../spec/test-vectors.json';
import { decodeChain, decodeEvmUnsigned, decodeSignableMessage } from './decoders.js';

const EVM_KINDS = new Set([
    'ethereum',
    'polygon',
    'bsc',
    'arbitrum',
    'optimism',
    'base',
    'customEvm',
]);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const all: any[] = (vectors as any).vectors;

describe('EVM address vectors', () => {
    let ran = 0;
    for (const v of all) {
        if (v.kind !== 'address') continue;
        if (!EVM_KINDS.has(v.input.chain.kind)) continue;

        it(`address: ${v.id}`, () => {
            const wallet = JovaWallet.fromMnemonic({
                words: v.input.mnemonic,
                passphrase: v.input.passphrase ?? '',
            });
            try {
                const got = wallet.address(decodeChain(v.input.chain), 0);
                expect(got.value.toLowerCase()).toBe(
                    (v.expected.address as string).toLowerCase(),
                );
            } finally {
                wallet.destroy();
            }
            ran++;
        });
    }

    it('exercised at least one EVM address vector', () => {
        expect(ran).toBeGreaterThan(0);
    });
});

describe('EVM sign_tx vectors', () => {
    let ran = 0;
    for (const v of all) {
        if (v.kind !== 'sign_tx') continue;
        if (v.input.unsigned_tx.kind !== 'evm') continue;

        it(`sign_tx: ${v.id}`, () => {
            const wallet = JovaWallet.fromMnemonic({
                words: v.input.mnemonic,
                passphrase: v.input.passphrase ?? '',
            });
            try {
                const unsigned = { kind: 'evm' as const, ...decodeEvmUnsigned(v.input.unsigned_tx) };
                const signed = wallet.signTx(unsigned);
                expect(signed.raw_hex.toLowerCase()).toBe(
                    (v.expected.signed_hex as string).toLowerCase(),
                );
                expect(signed.tx_hash.toLowerCase()).toBe(
                    (v.expected.tx_hash as string).toLowerCase(),
                );
            } finally {
                wallet.destroy();
            }
            ran++;
        });
    }

    it('exercised at least one EVM sign_tx vector', () => {
        expect(ran).toBeGreaterThan(0);
    });
});

describe('EVM sign_message vectors', () => {
    let ran = 0;
    for (const v of all) {
        if (v.kind !== 'sign_message') continue;
        const k = v.input.message.kind;
        if (k !== 'evmPersonalSign' && k !== 'evmTypedDataV4') continue;

        it(`sign_message: ${v.id}`, () => {
            const wallet = JovaWallet.fromMnemonic({
                words: v.input.mnemonic,
                passphrase: v.input.passphrase ?? '',
            });
            try {
                const msg = decodeSignableMessage(v.input.message);
                const sig = wallet.signMessage(msg);
                expect(sig.hex.toLowerCase()).toBe(
                    (v.expected.signature as string).toLowerCase(),
                );
            } finally {
                wallet.destroy();
            }
            ran++;
        });
    }

    it('exercised at least one EVM sign_message vector', () => {
        expect(ran).toBeGreaterThan(0);
    });
});

describe('EVM error vectors', () => {
    let ran = 0;
    for (const v of all) {
        if (v.kind !== 'error') continue;
        // Phase 1 error vectors go through sign_tx with a malformed tx.
        if (!v.input.unsigned_tx) continue;
        if (v.input.unsigned_tx.kind !== 'evm') continue;
        // ID convention: Phase 1 used `phase1.error.*`; Phase 6 keeps that
        // filter to avoid pulling in btc.error / sol.error / xrp.error
        // vectors which are scoped to their per-chain test files.
        if (!String(v.id).startsWith('phase1.error.')) continue;

        it(`error: ${v.id}`, () => {
            const wallet = JovaWallet.fromMnemonic({
                words: v.input.mnemonic,
                passphrase: v.input.passphrase ?? '',
            });
            try {
                const unsigned = { kind: 'evm' as const, ...decodeEvmUnsigned(v.input.unsigned_tx) };
                let caught: unknown = null;
                try {
                    wallet.signTx(unsigned);
                } catch (e) {
                    caught = e;
                }
                expect(caught).toBeInstanceOf(JovaException);
                // The Rust JovaError variant names map onto the WASM
                // JovaErrorPayload `kind` discriminator with camelCase
                // ("MalformedUnsignedTx" → "malformedUnsignedTx").
                const expected = (v.expected.error_variant as string);
                const expectedKind = expected.charAt(0).toLowerCase() + expected.slice(1);
                const payload = (caught as JovaException).error;
                expect(payload.kind).toBe(expectedKind);
            } finally {
                wallet.destroy();
            }
            ran++;
        });
    }

    it('exercised at least one EVM error vector', () => {
        expect(ran).toBeGreaterThan(0);
    });
});

File: bindings/kotlin/jova-core/src/test/kotlin/io/jova/core/EvmVectorsTest.kt
package io.jova.core

import org.junit.Test
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.json.JSONObject
import org.json.JSONArray
import uniffi.jova_core_ffi.FfiException
import uniffi.jova_core_ffi.JovaChain
import uniffi.jova_core_ffi.JovaWallet
import uniffi.jova_core_ffi.UnsignedTx

class EvmVectorsTest {

    private fun loadVectors(): JSONArray {
        val raw = javaClass.getResourceAsStream("/test-vectors.json")!!
            .bufferedReader().readText()
        return JSONObject(raw).getJSONArray("vectors")
    }

    @Test
    fun evmAddressVectors() {
        val vectors = loadVectors()
        var ran = 0
        for (i in 0 until vectors.length()) {
            val v = vectors.getJSONObject(i)
            if (v.getString("kind") != "address") continue
            val input = v.getJSONObject("input")
            val chain = decodeChain(input.getJSONObject("chain"))
            // EVM-only in this test; non-EVM chains get their own test in Phase 2/3.
            when (chain) {
                JovaChain.Ethereum, JovaChain.Polygon, JovaChain.Bsc,
                JovaChain.Arbitrum, JovaChain.Optimism, JovaChain.Base,
                is JovaChain.CustomEvm -> { /* keep */ }
                else -> continue
            }

            val mnemonic = input.getString("mnemonic")
            val pass = if (input.has("passphrase")) input.getString("passphrase") else ""
            val expected = v.getJSONObject("expected").getString("address")

            val wallet = JovaWallet.fromMnemonic(mnemonic, pass)
            val got = wallet.address(chain, 0u)
            assertEquals(
                "address mismatch for vector ${v.getString("id")}",
                expected.lowercase(),
                got.value.lowercase()
            )
            ran++
        }
        assertTrue("No address vectors ran — check vector file", ran > 0)
    }

    @Test
    fun evmSignTxVectors() {
        val vectors = loadVectors()
        var ran = 0
        for (i in 0 until vectors.length()) {
            val v = vectors.getJSONObject(i)
            if (v.getString("kind") != "sign_tx") continue
            val input = v.getJSONObject("input")
            val mnemonic = input.getString("mnemonic")
            val pass = if (input.has("passphrase")) input.getString("passphrase") else ""
            val unsignedDict = input.getJSONObject("unsigned_tx")

            if (unsignedDict.getString("kind") != "evm") continue
            val unsigned = UnsignedTx.Evm(tx = decodeEvmUnsigned(unsignedDict))
            val expected = v.getJSONObject("expected")

            val wallet = JovaWallet.fromMnemonic(mnemonic, pass)
            val signed = wallet.signTx(unsigned, 0u)
            assertEquals(
                "signed_hex mismatch for vector ${v.getString("id")}",
                expected.getString("signed_hex").lowercase(),
                signed.rawHex.lowercase()
            )
            assertEquals(
                "tx_hash mismatch for vector ${v.getString("id")}",
                expected.getString("tx_hash").lowercase(),
                signed.txHash.lowercase()
            )
            ran++
        }
        assertTrue("No sign_tx vectors ran — check vector file", ran > 0)
    }

    @Test
    fun evmSignMessageVectors() {
        val vectors = loadVectors()
        var ran = 0
        for (i in 0 until vectors.length()) {
            val v = vectors.getJSONObject(i)
            if (v.getString("kind") != "sign_message") continue
            val input = v.getJSONObject("input")
            val mnemonic = input.getString("mnemonic")
            val pass = if (input.has("passphrase")) input.getString("passphrase") else ""
            val messageDict = input.getJSONObject("message")

            // Phase 2 added BTC sign_message vectors with `message.kind == "bitcoin"`;
            // those are exercised by BtcVectorsTest. Skip them here. (The decoder
            // used to throw on non-EVM kinds; now it returns Bitcoin variants too.)
            val msgKind = messageDict.getString("kind")
            if (msgKind != "evmPersonalSign" && msgKind != "evmTypedDataV4") continue

            val signable = try {
                decodeSignableMessage(messageDict)
            } catch (e: VectorDecodeException) {
                continue
            }

            val expectedSig = v.getJSONObject("expected").getString("signature").lowercase()

            val wallet = JovaWallet.fromMnemonic(mnemonic, pass)
            val sig = wallet.signMessage(signable, 0u)
            assertEquals(
                "signature mismatch for vector ${v.getString("id")}",
                expectedSig,
                sig.hex.lowercase()
            )
            ran++
        }
        assertTrue("No sign_message vectors ran — check vector file", ran > 0)
    }

    @Test
    fun evmErrorVectors() {
        val vectors = loadVectors()
        var ran = 0
        for (i in 0 until vectors.length()) {
            val v = vectors.getJSONObject(i)
            if (v.getString("kind") != "error") continue
            val input = v.getJSONObject("input")
            val mnemonic = input.getString("mnemonic")
            val pass = if (input.has("passphrase")) input.getString("passphrase") else ""
            val expected = v.getJSONObject("expected")
            val expectedVariant = expected.getString("error_variant")

            // All Phase 1 error vectors go through sign_tx with a malformed tx.
            if (!input.has("unsigned_tx")) continue
            val unsignedDict = input.getJSONObject("unsigned_tx")
            if (unsignedDict.getString("kind") != "evm") continue

            val unsigned = UnsignedTx.Evm(tx = decodeEvmUnsigned(unsignedDict))
            val wallet = JovaWallet.fromMnemonic(mnemonic, pass)
            try {
                wallet.signTx(unsigned, 0u)
                fail("Expected FfiException.$expectedVariant for vector ${v.getString("id")}, but no exception was thrown")
            } catch (err: FfiException) {
                val actualVariant = err::class.simpleName ?: "Unknown"
                assertEquals(
                    "Wrong error variant for vector ${v.getString("id")}",
                    expectedVariant,
                    actualVariant
                )
                ran++
            }
        }
        assertTrue("No error vectors ran — check vector file", ran > 0)
    }
}
